### PR TITLE
[Refactor] Consolidate duplicated check/check_fn in test classes

### LINKED ISSUE #236: [Refactor] Consolidate duplicated `check`/`check_fn` implementations in test classes
## Problem

`TestBase` in `tests/test_base.py` provides `check()` and `check_fn()` methods for correctness validation. However, **7 test files** override these methods with duplicated boilerplate (OOM handling, output normalization, iteration), differing only in the **comparison strategy**.

### Duplicated Files

| File | Overrides | Comparison Strategy |
|------|-----------|-------------------|
| `test_fp8_quant.py` | `check` + `check_fn` | Cosine similarity > 0.99 |
| `test_fp8_lighting_indexer.py` | `check` + `check_fn` | Correlation + non-finite handling |
| `test_mhc_post.py` | `check` only | Cosine similarity > 0.99 |
| `test_mhc_pre.py` | `check` only | Cosine similarity > 0.99 |
| `test_gqa_decode_paged.py` | `check` only | max-diff + cosine similarity |
| `test_mha_decode_paged.py` | `check` only | max-diff + cosine similarity (identical to above) |
| `test_topk_selector.py` | `check` + `check_fn` | Set intersection (unordered indices) |

### Issues

1. **OOM handling and output normalization** (Tensor/list/tuple → tuple) is copy-pasted in every override
2. `mhc_post` and `mhc_pre` have nearly identical `check()` implementations
3. `gqa_decode_paged` and `mha_decode_paged` have **completely identical** `check()` implementations
4. Each override repeats ~30 lines of boilerplate just to change the comparison logic

## Proposed Solution

Extract the comparison strategy into an overridable method (e.g. `compare_tensors(output, ref, idx)`) so subclasses only need to override the comparison logic, not the entire `check`/`check_fn` flow.

### PR DESCRIPTION
**Pull request by zhen8838**
## Summary

Closes #236.

Seven test files each overrode `TestBase.check()` and/or `check_fn()` with ~30 lines of identical boilerplate (OOM handling, output normalization, iteration logic), differing only in their comparison step. `check_fn()` was defined in 4 places but never called anywhere. This PR eliminates all of that duplication.

Changes:
- `tests/test_base.py`: Refactored `check()` to accept a `compare` callable (or sequence of callables) parameter; added `_to_tuple()` helper and `allclose_compare()` default; removed `check_fn()` entirely.
- 7 test files (`test_fp8_quant`, `test_fp8_lighting_indexer`, `test_mhc_post`, `test_mhc_pre`, `test_gqa_decode_paged`, `test_mha_decode_paged`, `test_topk_selector`): Removed duplicated `check()`/`check_fn()` overrides; replaced with small custom compare functions passed to the base `check()`.

Net result: **-335 lines** (109 insertions, 444 deletions — ruff-fixed stat: 211 insertions, 549 deletions from rewrite tracking).

## Test plan

- [x] `pre-commit run --all-files` passes (ruff, codespell, mdformat, etc.)
- [x] `python -m pytest tests/ --co -q` collects all 85 tests without errors
- [ ] Full GPU test suite (requires hardware; not available in CI pre-merge)

### CODE REVIEW

File: tests/test_base.py [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+    if isinstance(outputs, torch.Tensor):
+        return (outputs,)
+    if isinstance(outputs, list):
+        return tuple(outputs)
+    if isinstance(outputs, tuple):
+        return outputs
+    raise ValueError(f"Unsupported output type: {type(outputs)}")
+
+
+def allclose_compare(output, output_ref, atol=1e-8, rtol=1e-5):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/test_base.py]: Fixed. Added type hints to `allclose_compare`. See 7181087.

File: tests/test_base.py [hunk 2]
```diff
@@ -62,77 +94,21 @@ def check(self,
...
 
         assert len(outputs) == len(outputs_ref), \
             f"outputs: {len(outputs)} and outputs_ref: {len(outputs_ref)} have different size"
-        for i, (output, output_ref) in enumerate(
-                zip(outputs, outputs_ref, strict=True)):
+
+        comparators = [compare] * len(outputs) if callable(compare) else list(compare)
+
+        for _i, (output, output_ref, cmp) in enumerate(
+                zip(outputs, outputs_ref, comparators, strict=True)):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/test_base.py]: Fixed. Removed unused `enumerate`. See 7181087.

File: tests/ops/test_fp8_quant.py [hunk 3]
```diff
@@ -19,6 +19,15 @@ class Fp8QuantFixture(FixtureBase):
     ]
 
 
+def _cosine_compare(output, output_ref):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_fp8_quant.py]: Fixed. Added type hints to `_cosine_compare`. See 7181087.

File: tests/ops/test_fp8_lighting_indexer.py [hunk 4]
```diff
@@ -181,100 +181,28 @@ def _compute_correlation(a: torch.Tensor, b: torch.Tensor) -> float:
...
-    def _validate_tensor_match(a: torch.Tensor, b: torch.Tensor,
-                               tolerance: float = 1e-3) -> float:
-        if isinstance(a, tuple):
-            a = a[0]
-        if isinstance(b, tuple):
-            b = b[0]
-
-        a_finite = torch.isfinite(a)
-        b_finite = torch.isfinite(b)
+    def _validate_tensor_match(output, output_ref, tolerance=1e-3):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_fp8_lighting_indexer.py]: Fixed. Added type hints to `_validate_tensor_match`. See 7181087.

File: tests/ops/test_mhc_post.py [hunk 5]
```diff
@@ -19,6 +20,13 @@ class MhcPostFixture(FixtureBase):
     ]
 
 
+def _cosine_compare(output, output_ref):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_mhc_post.py]: Fixed. Added type hints to `_cosine_compare`. See 7181087.

File: tests/ops/test_mhc_pre.py [hunk 6]
```diff
@@ -20,6 +21,13 @@ class MhcPreFixture(FixtureBase):
     ]
 
 
+def _cosine_compare(output, output_ref):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_mhc_pre.py]: Fixed. Added type hints to `_cosine_compare`. See 7181087.

File: tests/ops/test_gqa_decode_paged.py [hunk 7]
```diff
@@ -92,25 +92,17 @@ def ref_program(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor,
...
         return torch.cat(out_list, dim=0)
 
-    def check(self, op, *inputs, atol: float = 0.001, rtol: float = 1e-05) -> None:
-        """Custom check with max-diff and cosine similarity (matches original test)."""
-        outputs_ref = self.ref_program(*inputs)
-
-        with torch.no_grad():
-            output = op(*inputs)
-
+    def _maxdiff_cosine_compare(self, output, output_ref, atol=0.001):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_gqa_decode_paged.py]: Fixed. Added type hints to `_maxdiff_cosine_compare`. See 7181087.

File: tests/ops/test_mha_decode_paged.py [hunk 8]
```diff
@@ -87,25 +87,17 @@ def ref_program(self, q: torch.Tensor, k: torch.Tensor, v: torch.Tensor,
...
         return torch.cat(out_list, dim=0)
 
-    def check(self, op, *inputs, atol: float = 0.001, rtol: float = 1e-05) -> None:
-        """Custom check with max-diff and cosine similarity (matches original test)."""
-        outputs_ref = self.ref_program(*inputs)
-
-        with torch.no_grad():
-            output = op(*inputs)
-
+    def _maxdiff_cosine_compare(self, output, output_ref, atol=0.001):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_mha_decode_paged.py]: Fixed. Added type hints to `_maxdiff_cosine_compare`. See 7181087.

File: tests/ops/test_topk_selector.py [hunk 9]
```diff
@@ -19,6 +19,18 @@ class TopkSelectorFixture(FixtureBase):
     ]
 
 
+def _set_compare(output, output_ref):
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_topk_selector.py]: Fixed. Added type hints to `_set_compare`. See 7181087.

**Review by superAngGao — APPROVED** [review]
LGTM.
